src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/FaultDetectionStatsCollector.java

```diff
@@ -0,0 +1,253 @@
+                    mapper.writeValueAsString(getFollowerCheckStats()), FaultDetectionStats.class);
+            FaultDetectionStats leaderCheckStats = mapper.readValue(
+                    mapper.writeValueAsString(getLeaderCheckStats()), FaultDetectionStats.class);
+
+            FaultDetectionMetrics faultDetectionMetrics = new FaultDetectionMetrics();
+            if (followerCheckStats != null) {
+                faultDetectionMetrics.setFollowerCheckMetrics(
+                        computeLatency(followerCheckStats, FaultDetectionStatsCollector.prevFollowerCheckStats),
+                        computeFailure(followerCheckStats, FaultDetectionStatsCollector.prevFollowerCheckStats));
+                FaultDetectionStatsCollector.prevFollowerCheckStats = followerCheckStats;
+            }
+            if (leaderCheckStats != null) {
+                faultDetectionMetrics.setLeaderCheckMetrics(
+                        computeLatency(leaderCheckStats, FaultDetectionStatsCollector.prevLeaderCheckStats),
+                        computeFailure(leaderCheckStats, FaultDetectionStatsCollector.prevLeaderCheckStats));
+                FaultDetectionStatsCollector.prevLeaderCheckStats = leaderCheckStats;
+            }
+
+            StringBuilder value = new StringBuilder();
+            value.append(PerformanceAnalyzerMetrics.getJsonCurrentMilliSeconds())
+                    .append(PerformanceAnalyzerMetrics.sMetricNewLineDelimitor);
+            value.append(faultDetectionMetrics.serialize());
+            saveMetricValues(value.toString(), startTime);
+
+            PerformanceAnalyzerApp.WRITER_METRICS_AGGREGATOR.updateStat(
+                    WriterMetrics.FAULT_DETECTION_COLLECTOR_EXECUTION_TIME, "",
+                    System.currentTimeMillis() - mCurrT);
+        } catch (NoSuchFieldException | InvocationTargetException | IllegalAccessException | NoSuchMethodException
+                | JsonProcessingException ex) {
+            PerformanceAnalyzerApp.ERRORS_AND_EXCEPTIONS_AGGREGATOR.updateStat(
+                    ExceptionsAndErrors.FAULT_DETECTION_COLLECTOR_ERROR, "", 1);
+            LOG.debug("Exception in Collecting Fault Detection Stats: {} for startTime {}",
+                    () -> ex.toString(), () -> startTime);
+        }
+    }
+
+    @VisibleForTesting
+    public Object getLeaderCheckStats() throws InvocationTargetException, IllegalAccessException,
+            NoSuchMethodException, NoSuchFieldException {
+        Method method = LeaderChecker.class.getMethod(GET_STATS_METHOD_NAME);
+        Discovery discovery = ESResources.INSTANCE.getDiscovery();
+        if(discovery instanceof Coordinator) {
+            Coordinator coordinator = (Coordinator) discovery;
+            Field leaderCheckerField = Coordinator.class.getDeclaredField(LEADER_CHECKER_FIELD);
+            leaderCheckerField.setAccessible(true);
+            LeaderChecker leaderChecker = (LeaderChecker) leaderCheckerField.get(coordinator);
```

Comment:
Can this reflection code be refactored outside the collectMetrics so that it is run only once?